### ISSUE TITLE
Spurious th element in HTML table body

### ISSUE THREAD

**Issue by dginev** [brucemiller/LaTeXML#591]
The LaTeX source:
```latex
\begin{tabular}{ll}
H1 & H2\\   
A & 0\\

\end{tabular}
```

When run with ```latexmlc --format=html5``` produces:
```html
<table xmlns="http://www.w3.org/1999/xhtml" id="p1.1" class="ltx_tabular ltx_align_middle">
<thead class="ltx_thead">
<tr id="p1.1.1" class="ltx_tr">
<th id="p1.1.1.1" class="ltx_td ltx_align_left">H1</th>
<th id="p1.1.1.2" class="ltx_td ltx_align_left">H2</th></tr>
</thead>
<tbody class="ltx_tbody">
<tr id="p1.1.2" class="ltx_tr">
<th id="p1.1.2.1" class="ltx_td ltx_align_left">A</th>
<td id="p1.1.2.2" class="ltx_td ltx_align_left">0</td></tr>
<tr id="p1.1.3" class="ltx_tr">
<td id="p1.1.3.1" class="ltx_td ltx_align_left"/>
<td id="p1.1.3.2" class="ltx_td"/></tr>
</tbody>
</table>
```

[Click here](http://latexml.mathweb.org/editor?tex=%5Cbegin%7Btabular%7D%7Bll%7D%0AH1%20%26%20H2%5C%5C%20%20%20%0AA%20%26%200%5C%5C%0A%0A%5Cend%7Btabular%7D%0A) for a live example

Note that inside the table rows inside the table body, the first column is always a ```th``` instead of a ```td```. I can only assume this is a bug, though I am currently a bit mystified about what might be causing it. Removing the last empty line in the table resolves the problem, so it may be related.

**Comment by brucemiller** [comment]
Well, it's not quite the bug that you think it is. It's looking at apparent distinctions between the formatting and contents of different rows and columns to attempt to infer whether the row or column is likely to be a heading.  Often the 1st column of a table is a heading.  No idea, without digging, why the blank row should affect it here.  The method is wildly heuristic and while more often right than wrong, it's hardly 100%

**Comment by dginev** [comment]
We should coin the term HI for "Heuristic Intelligence". Can't we have a special "ltx_guessed_th" class to mark those? Then I can style based on how much I believe the heuristic :>

**Comment by brucemiller** [comment]
well, *every* <th> in a latexml table is guessed. But I'm surprised that it didn't get some unique class anyway, just ltx_td.

**Comment by dginev** [comment]
Ok, got around to reading the details of the Alignment heuristics. Very interesting!

I am going to reframe this issue as a discussion, rather than a bug report, now that I understand more. 
 1. Do I understand well, that if one wants to mimic the LaTeX styling in HTML, there should be no CSS rules based on ```td``` and ```th```, instead all of them should be about ```.ltx_td```?
 2. I think when LaTeXML guesses a column header it should also add the scope attribute as an annotation: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/th#attr-scope
 3. Would the size of the table be useful for the heuristics? E.g. a table with 2 columns is harder to guess than a table with 10 columns. (1 column of text with 9 cols of numbers is a lot more convincing than 1-to-1).

I probably can't request any meaningful "fix" for the example I gave.

**Comment by brucemiller** [comment]
Interesting; didn't know about the scope attribute. It probably could be inferred from the heuristics --- or at least, whether LaTeXML *thinks* it's a row/column heading!  OTOH, scope only applies to th. And I was going to say that it is presumptuous to even use th instead of td; that it would be better to only add a ltx_guessed_th to class!

I think that the size of the table is indirectly affecting the heurstic, but would be hard put to convince you (or me).

**Comment by dginev** [comment]
Right, if you are guessing styling, it would be awesome if that is always captured by a special CSS class, even if LaTeXML also produces special element markup. Being able to quickly write CSS selectors for particular heuristic markup is a super powerful customization tool.

In fact, I am missing that right now, as I am embedding the LaTeXML document in a larger context which could have its own HTML tables and th elements, and I need to write complex selectors to only target the LaTeXML ones. It is of course possible, so even if we close here with a wontfix, that is perfectly ok.

**Comment by brucemiller** [comment]
Not quite obvious how to propogate the, otherwise useless, scope attribute up to the html (and that's presumably only an html5 attribute?).  How about having distinct row/column class values?  ltx_guessed_row_th, ltx_guessed_col_th ?

**Comment by dginev** [comment]
It's css, let's have several:
 * ```ltx_th ltx_row_th```
 * ```ltx_th ltx_col_th```

If all of them are guessed, no harm in dropping the ```guessed``` bit, it's cleaner without it.

**Comment by brucemiller** [comment]
Sure, but you'd asked about deluxetable; I took a brief look at it, and it actually (and nicely) allows you to declare both column and row headers!  So, in that case, they presumably would no-longer be guessed.  (you can turn off the guessing, btw.... somehow).  Is that a distinction that's worth preserving?

**Comment by dginev** [comment]
Oh nice. Yes, then it is worth preserving, since someone may not like the guessing for column headers and open an issue about it :> Then I suggest:
 * ``` ltx_th ltx_guessed_th ltx_row_th```
 * ``` ltx_th ltx_guessed_th ltx_col_th```

For the guessed headers. More classes means finer granularity for the CSS hooks.

**Comment by brucemiller** [comment]
I almost missed my opportunity to (legitimately, for once) say "That's not a bug, that's a *feature*!"

**Comment by brucemiller** [comment]
Ok, so now we get class ```ltx_th, ltx_th_row, ltx_th_column``` on th elements, depending on whether they've been marked as being row or column headings. Since html5 now allows tfoot to appear at the end of tables, we also wrap any trailing rows containing only th with tfoot, just as leading ones are wrapped in thead. 

When LaTeXML has guessed at headings, the table gets class ```ltx_guessed_headers```. Alternatively, with latexml.sty you can use ```\lxBeginTableHead, \lxEndTableHead, \lxBeginTableFoot, \lxEndTableFoot, \lxTableColumnHead, \lxTableRowHead``` in a somewhat obvious fashion to mark row & column headers explicltly.

**Comment by dginev** [comment]
That sounds perfect! Have you also updated the LaTeXML CSS files?

**Comment by brucemiller** [comment]
yep, although by default there's no particular styling of headers...

**Comment by dginev** [comment]
Just to finalize the discussion here - the "th" bold was provided by my own LaTeXML-external CSS, as we consider it a rule of thumb to have table headers bold. I added a special rule:
```
tbody th {
  font-weight: normal;
}
```

to avoid the bold effect inside table bodies. I'm wondering whether you shouldn't be using ```td``` elements instead of ```th``` inside table bodies, and then specify them using the classes. But one way or another, I have resolved the problem.
